### PR TITLE
feat: open squads page adjustment

### PR DESCRIPTION
**Pull request by capJavert**
### Describe what this PR does
- Adjustments for open squad page 
- This also makes open squad page visible to anonymous users
- added `useJoinReferral` hook to be able to handle setting referral cookie in different places on the client
- added `useJoinSquad` hook and refactored `useLeaveSquad` to only contain logic without side effects
  - with that both hook functions can be used with `useMutation` to support loading and all other cases where needed
- needs https://github.com/dailydotdev/daily-api/pull/1341 for API referralUrl to work

## Events

Did you introduce any new tracking events?
Don't forget to update the [Analytics Taxonomy sheet](https://docs.google.com/spreadsheets/d/18Lv7zXges9QfVX5VYL1a-Hyl0e1sQ3sLr0OK8YZWKXI/edit#gid=0)

Log the new/changed events below:

| Type | event_name | target_type | value |
|--------|-------|------|-------|
|New | impression | squad join button | extra: { squad, origin, squad_type } |

### On those affected packages:
- [ ] Have you done sanity checks in the webapp?
- [ ] Have you done sanity checks in the extension?
- [ ] Does this not break anything in companion?

### Did you test the modified components media queries?
- [ ] MobileL (420px)
- [ ] Tablet (656px)
- [ ] Laptop (1020px)

#### Did you test on actual mobile devices?
- [ ] iOS (Chrome and Safari)
- [ ] Android

WT-{number} #done

### CODE REVIEW

File: packages/shared/src/components/squads/SharePostBar.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SharePostBar.tsx:76]: This is the disabled render only right, not sure why this change was needed then?

File: packages/shared/src/components/squads/SquadHeaderMenu.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadHeaderMenu.tsx:103]:
```suggestion
        onClick: onLeaveSquad,
```

This should do right

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadJoinButton.tsx:31]: Please check Tom's latest PR it introduces labels for re-usable texts I see some duplicates here as well, so loved to start using it.

File: packages/shared/src/hooks/useJoinReferral.ts
Comment (rebelchris) [inline packages/shared/src/hooks/useJoinReferral.ts:33]: Should we not consider adding a removeCookie function in the cookie lib?
Could take a name for instance and do everything else inside?

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/[token].tsx:132]: The token here is either a user token or the squad token?
Just for my understanding.

**Review by rebelchris — CHANGES_REQUESTED** [review]
Nice work, some minor suggestions and a question.

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (capJavert) [reply packages/webapp/pages/squads/[handle]/[token].tsx:132]: Here it is always user token, with open squad we don't send token at all.

Referral is attributed through new referral cookie.

File: packages/shared/src/components/squads/SquadHeaderMenu.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadHeaderMenu.tsx:103]: Typescript complaining.

File: packages/shared/src/components/squads/SharePostBar.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SharePostBar.tsx:76]: This copy is from open squad page design.

Comment (rebelchris) [reply packages/shared/src/components/squads/SharePostBar.tsx:76]: But do they not intertwine?
Just making sure the evaluation is ok because the setting to even render this might be different.

File: packages/shared/src/components/squads/SquadHeaderBar.tsx
Comment (capJavert) [inline packages/shared/src/components/squads/SquadHeaderBar.tsx:61]: https://dailydotdev.slack.com/archives/C0592NL7YCU/p1689595125220649

File: packages/shared/src/hooks/useJoinReferral.ts
Comment (capJavert) [reply packages/shared/src/hooks/useJoinReferral.ts]: Added as `expireCookie` bb675b5c824b20bae4f2a26fd63c1ef9884b8d07

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:31]: Waiting for https://github.com/dailydotdev/apps/pull/1941 to merge and I'll add

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadJoinButton.tsx:30]: Does squad not have a `currentMember` property already?

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadJoinButton.tsx:32]: Is the useJoinSquad not already a mutation?
I think ideally that logic all sits in the hook normally

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadJoinButton.tsx:41]: Basically same comment for all the hooks

File: packages/shared/src/components/squads/SquadPageHeader.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadPageHeader.tsx:129]: Would still not be valid if the public squad would be set to only admins right?
Just checking for future reference, I know it's an edge-case for now

File: packages/shared/src/hooks/useJoinReferral.ts
Comment (rebelchris) [inline packages/shared/src/hooks/useJoinReferral.ts:21]: This can be top level outside component

File: packages/shared/src/hooks/useJoinReferral.ts
Comment (rebelchris) [inline packages/shared/src/hooks/useJoinReferral.ts:18]: Optional, but maybe open for discussion...

I personally find it more readable to have a function defined that handles the logic and the useQuery to simply invoke that function.
Would love to hear if we can generalise one way moving forward.

File: packages/shared/src/hooks/useSquadOnboarding.ts
Comment (rebelchris) [inline packages/shared/src/hooks/useSquadOnboarding.ts:30]: Just checking do we not have something like isLoggedUser?
Maybe more explicit moving forward.

File: packages/webapp/pages/join.tsx
Comment (rebelchris) [inline packages/webapp/pages/join.tsx:68]: I see multiple introductions of this, maybe we can util it :)

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/[token].tsx:139]: Did we intentionally remove the tracking here?

File: packages/webapp/pages/squads/[handle]/index.tsx
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/index.tsx:152]: We also have `ConditionalWrapper` that can handle these kind of things, leaving it up to you what to use though.

**Review by rebelchris — CHANGES_REQUESTED** [review]
Some small comments, nothing major looks pretty good!

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (capJavert) [reply packages/webapp/pages/squads/[handle]/[token].tsx:139]: No it is moved to `useJoinSquad`

File: packages/shared/src/hooks/useSquadOnboarding.ts
Comment (capJavert) [reply packages/shared/src/hooks/useSquadOnboarding.ts:30]: No, I think actually it might be good to add it to auth context.

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:32]: Not really it is just a function that is exported that does "bunch of things that are related". Kinda like your comment https://github.com/dailydotdev/apps/pull/1948#discussion_r1266810470 with which I agree.

If we start just re-exporting useMutation wrapped in custom hook we then often return from it something like:
`{ onJoinSquad, isLoading }`

Then we need for example another property `mutation.isDone`, we add that.

Then we need `onSuccess`, we add that to hook props.. and so on..

Pretty fast hook has X props and Y return values.

I think it is much cleaner and we should think of `useMutation` as a wrapper for loading state, success/error callbacks in specific implementation vs hardcoding them inside custom hooks which then become really specific and hard to expand.

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx]: It was because squad query would get stale due to only depending on handle and `currentMember` would be left false. Fixed here 06aa0932e015cb18fbc12d3ed37ce8f7625fe01a

File: packages/shared/src/hooks/squads/useSquad.ts
Comment (capJavert) [inline packages/shared/src/hooks/squads/useSquad.ts:23]: invalidateQueries with `['squad', handle]` still work because we use array query keys.

File: packages/webapp/pages/squads/[handle]/index.tsx
Comment (capJavert) [reply packages/webapp/pages/squads/[handle]/index.tsx:152]: Hmm yeah, but would not work since `ProtectedPage` requires `seo`, and I would have to change the interface to support it and then again move more things around. Would prefer keeping it here as current.

File: packages/shared/src/components/squads/SquadPageHeader.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadPageHeader.tsx:129]: It would not be valid only if user was not a member, since it would always say `Join the squad to create new posts`. But I think that is copy wise because:
1. we might wanna always invite people to join (no matter of squad preferences)
2. we don't expose actual posting permission to anon users (so we could not calculate it)

We would need API adjustments to allow `memberPostingRank` for anon users and check with product about the messaging. I personally think it would just take time but I can pursue if you want.

File: packages/shared/src/hooks/useSquadOnboarding.ts
Comment (capJavert) [reply packages/shared/src/hooks/useSquadOnboarding.ts:30]: I can add in `AuthContext`?

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:29]: Alright valid point, but at which you would almost argue why the custom hook then.
For me it's a wrapping place where all related logic happens so we don't have to do it in multiple places.

Tracking, rollbacks, toast etc

File: packages/shared/src/components/squads/SquadPageHeader.tsx
Comment (rebelchris) [reply packages/shared/src/components/squads/SquadPageHeader.tsx:129]: Alright let's leave it as is 👍

File: packages/shared/src/hooks/useSquadOnboarding.ts
Comment (rebelchris) [reply packages/shared/src/hooks/useSquadOnboarding.ts:30]: Would be a cool addition I think 👍
Seeing more and more of those cases

File: packages/webapp/pages/squads/[handle]/index.tsx
Comment (rebelchris) [reply packages/webapp/pages/squads/[handle]/index.tsx:152]: No problem 🙏

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadJoinButton.tsx:45]: Do we actually need to invalidate, what's the catch of not doing it?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Also seems we are missing the impression on the join button.
Would suggest to introduce it here in case we forget.
And see you already have the completion introduced here.

Screenshot 2023-07-19 at 13 14 36

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:45]: So that this is updated, it is optional though yes. Don't have the especial preference here if its worth it, WDYT?

image

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:29]: > Alright valid point, but at which you would almost argue why the custom hook then.

Custom hook here is because we need context and React context to actually get to our tracking function and updating boot. If not we would not need it and it could just be exported.

> Tracking, rollbacks, toast etc

I would argue side effects should not be there because more times those differ from component to component. 

Or we would have to be very specific for what hook is used. Basically each new argument we add to the hook we make it more complex and with more variations on how it works. 

We could thought copy what we are doing `useVotePost` where `mutationHandlers` are exported and reused across different usages of the hooks. 

That way we could still have the function combine all the related logic but side effects can be injected again where needed from predefined functions (or hooks).

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:45]: Ah I see, well let's do it.
Doesn't cost too much, was just wondering what actually would change 😅

**Comment by capJavert** [comment]
@rebelchris good catch, missed that new impression that was added to the table.

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [inline packages/shared/src/components/squads/SquadJoinButton.tsx:46]: Not a big fan of passing depedency for the sake of impression calls.
It's important they don't trigger more then once.

I'd either suggest simply removing dependencies here or adding a state to ensure it only fires once per mount.

I know very unlikely it would change, but keep in mind you invalidate queries, so we could accidentally trigger again.

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (rebelchris) [inline packages/webapp/pages/squads/[handle]/[token].tsx:139]: This should not live here right?
It's already handled inside the button which makes more sense.

Comment (rebelchris) [reply packages/webapp/pages/squads/[handle]/[token].tsx:139]: I see I think this is what I meant with duplicate code issue with splitting these buttons.
Can we find a way to uniform the behaviour?

It's pretty dangerous to have duplicated tracking events that do the exact same.
I'm not sure how much other stuff of this is duplicated?

**Review by rebelchris — CHANGES_REQUESTED** [review]
Just some concerns about the tracking.

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (idoshamun) [reply packages/webapp/pages/squads/[handle]/[token].tsx:139]: I agree this code duplication is highly error-prone. Ideally, we want to have one function/hook that triggers events. Duplicating is so risky, especially when it comes to tracking. I would suggest creating a hook to make sure that or a function. Happy with either of them

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:51]: But we need those dependencies, if squad id changes or button get's rendered with different squad ID or place (with different origin) we want to trigger impression again right? @rebelchris

Comment (rebelchris) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:51]: Which would be a new mount, afaik there is only a danger of retriggering impressions.
CC: @idoshamun Maybe your thoughts?

Comment (idoshamun) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:51]: If there's a chance that it will get rendered then it's better to have the deps. Otherwise let's stick to no deps. I think when we discussed with tkdodo he also mentioned tracking as one exception use case for not using the proper deps

File: packages/webapp/pages/squads/[handle]/[token].tsx
Comment (capJavert) [reply packages/webapp/pages/squads/[handle]/[token].tsx]: Yeah agreed, made adjustments here a798c15530f4168858f031567b739efea55d54d7, decided to move tracking logic to button which is now reused on squad invitation page and open squad page.

Changes:
- added `SimpleSquadJoinButton` which contains base logic and tracking for button
- this one is used for `SquadJoinButton` which is the new one with join/leave capabilities needed for open squad
- `SimpleSquadJoinButton` is used on squad invite token page
- refactored layout of squad token page to not render button twice on mobile/desktop (would trigger events twice) - visually layout is the same

**Comment by idoshamun** [comment]
@capJavert don't wait for my approval :)
I'm just jumping randomly on some comments and then disappear 😂

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (rebelchris) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:48]: @capJavert for me this one is still open 🫣

**Review by rebelchris — CHANGES_REQUESTED** [review]
Just the one thing, from my perspective just ensure it can only fire once, not for instance once on loading when one param is null, and then again once its set.

I really think empty props is ok for tracking for now and then we good to go.

File: packages/shared/src/components/squads/SquadJoinButton.tsx
Comment (capJavert) [reply packages/shared/src/components/squads/SquadJoinButton.tsx:48]: I'll just concede for now, let's discuss when we get on implementing special tracking hooks. 🤝

**Review by rebelchris — APPROVED** [review]
Looks good lets go 🙏

**Pull request merged**
